### PR TITLE
CI: Node 14 => Node 16

### PR DESCRIPTION
**Pull request by RobbieMcKinstry**
# Description

This PR migrates our CI/CD to use Node 16 instead of Node 14.
Node 14 will be sunset on April 30, 2023 (less than 3 weeks from the time of writing). This PR moves CI/CD to test and build on Node 16.

Fixes https://github.com/pulumi/pulumi/issues/12646

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by justinvp** [comment]
Do we also need to update the min version we test against?

https://github.com/pulumi/pulumi/blob/53e2a8b9a719eaf9589d86fed45c011144cbcdbc/scripts/get-job-matrix.py#L119

**Comment by RobbieMcKinstry** [comment]
> Do we also need to update the min version we test against?

Yes! I grepped around the codebase but must have missed this one!

**Comment by justinvp** [comment]
Took another look around the repo and found a couple other places:

https://github.com/pulumi/pulumi/blob/045ca5817652a40cd8542ab07e038ecfd8095e43/README.md?plain=1#L175-L176

https://github.com/pulumi/pulumi/blob/045ca5817652a40cd8542ab07e038ecfd8095e43/CONTRIBUTING.md?plain=1#L33

**Comment by RobbieMcKinstry** [comment]
We should also make an accompanying change to the docs.

**Comment by RobbieMcKinstry** [comment]
I think we want to wait until April 30 to merge this.

**Review by Frassle — APPROVED** [review]

**Pull request merged**
